internal/trie/node/decode.go

```diff
@@ -9,63 +9,60 @@ import (
 	"fmt"
 	"io"
 
-	"github.com/ChainSafe/gossamer/internal/trie/pools"
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
 var (
-	ErrReadHeaderByte     = errors.New("cannot read header byte")
-	ErrUnknownNodeType    = errors.New("unknown node type")
 	ErrDecodeValue        = errors.New("cannot decode value")
 	ErrReadChildrenBitmap = errors.New("cannot read children bitmap")
 	ErrDecodeChildHash    = errors.New("cannot decode child hash")
 )
 
 // Decode decodes a node from a reader.
+// The encoding format is documented in the README.md
+// of this package, and specified in the Polkadot spec at
+// https://spec.polkadot.network/#sect-state-storage
 // For branch decoding, see the comments on decodeBranch.
 // For leaf decoding, see the comments on decodeLeaf.
 func Decode(reader io.Reader) (n *Node, err error) {
-	buffer := pools.SingleByteBuffers.Get().(*bytes.Buffer)
-	defer pools.SingleByteBuffers.Put(buffer)
-	oneByteBuf := buffer.Bytes()
-	_, err = reader.Read(oneByteBuf)
+	variant, partialKeyLength, err := decodeHeader(reader)
 	if err != nil {
-		return nil, fmt.Errorf("%w: %s", ErrReadHeaderByte, err)
+		return nil, fmt.Errorf("cannot decode header: %w", err)
 	}
-	header := oneByteBuf[0]
 
-	nodeTypeHeaderByte := header >> 6
-	switch nodeTypeHeaderByte {
-	case leafHeader:
-		n, err = decodeLeaf(reader, header)
+	switch variant {
+	case leafVariant.bits:
+		n, err = decodeLeaf(reader, partialKeyLength)
 		if err != nil {
 			return nil, fmt.Errorf("cannot decode leaf: %w", err)
 		}
 		return n, nil
-	case branchHeader, branchWithValueHeader:
-		n, err = decodeBranch(reader, header)
+	case branchVariant.bits, branchWithValueVariant.bits:
+		n, err = decodeBranch(reader, variant, partialKeyLength)
 		if err != nil {
 			return nil, fmt.Errorf("cannot decode branch: %w", err)
 		}
 		return n, nil
 	default:
-		return nil, fmt.Errorf("%w: %d", ErrUnknownNodeType, nodeTypeHeaderByte)
+		// this is a programming error, an unknown node variant
+		// should be caught by decodeHeader.
+		panic(fmt.Sprintf("not implemented for node variant %08b", variant))
```

Comment:
No `decodeHeader` would return an error if it's unknown variant, and this function would return before the switch statement.

The panic is just for future development (what I'm currently doing) when adding variants to avoid going crazy. I think it's appropriate to use a panic when it can only be a programming error and not something coming from IO (files, OS, network, etc.)